source/fab/builder.py

```diff
@@ -74,12 +78,48 @@ def __init__(self, config: Config):
             (config.workspace / BUILD_OUTPUT).mkdir()
 
     def run(self):
+        self.init_logging()
+        self.init_metrics()
+
+        artefacts = dict()
+        try:
+            with TimerLogger(f'running {self.config.label} build steps') as steps_timer:
+                for step in self.config.steps:
+                    with TimerLogger(step.name) as step_timer:
+                        step.run(artefacts=artefacts, config=self.config)
+                    send_metric('steps', step.name, step_timer.taken)
+        except Exception as err:
+            raise Exception(f'\n\nError running build steps:\n{err}')
+        finally:
+            self.finalise_metrics(steps_timer.taken)
+
+    def init_logging(self):
         logger.info(f"{datetime.now()}")
-        logger.info(f"use_multiprocessing = {self.config.use_multiprocessing}")
         if self.config.use_multiprocessing:
-            logger.info(f"n_procs = {self.config.n_procs}")
+            logger.info(f'machine cores: {multiprocessing.cpu_count()}')
+            logger.info(f'available cores: {len(os.sched_getaffinity(0))}')
+            logger.info(f'using n_procs = {self.config.n_procs}')
+        logger.info(f"workspace is {self.config.workspace}")
 
-        artefacts = dict()
-        for step in self.config.steps:
-            with time_logger(step.name):
-                step.run(artefacts, self.config)
+    def init_metrics(self):
+        init_metrics()
+
+        send_metric(group='run', name='label', value=self.config.label)
+        send_metric(group='run', name='datetime', value=datetime.now().replace(microsecond=0).isoformat())
+        send_metric(group='run', name='sysname', value=os.uname().sysname)
+        send_metric(group='run', name='nodename', value=os.uname().nodename)
+        send_metric(group='run', name='machine', value=os.uname().machine)
+        send_metric(group='run', name='user', value=getpass.getuser())
+
+        if self.config.use_multiprocessing:
+            send_metric(group='run', name='machine cores', value=multiprocessing.cpu_count())
+            send_metric(group='run', name='available cores', value=len(os.sched_getaffinity(0)))
+            send_metric(group='run', name='using n_procs', value=self.config.n_procs)
```

Comment:
None of these seem like metrics. They seem more like technical diagnostics. Potentially their own logging stream. 

Just seeing the name "metrics" makes me think of time series data which none of this is.

Is this a problem with the name not reflecting the purpose or with overloading something intended for something else?

Reply:
Some metrics only become a time series over multiple runs. The classic example is the response time of a server's endpoint. You'd record that single metric every run and analyse the trend over time. This is where a tool like Prometheus could be useful as it can [do all that](https://prometheus.io/docs/visualization/grafana/) for us.

I think you're right to question whether the number of cores, etc, is useful in metrics. The reason they're there right now is merely for annotating the charts. Like logging, I expect we'll be tweaking what metrics are recorded as development progresses. In the future we might want to visualise how run time is affected by the number of utilised cores, which would presumably make this a useful metric but it seems too early to speculate. Always happy to chat further in case I've missed something.

Reply:
As long as there is a plan which includes reassessing current usage I am okay with it as is.